File: Backend_FastAPI/app/schemas/system_config.py
# app/schemas/system_config.py
"""SystemConfig Pydantic schemas (#184 Wave 1 PR-1D / phase1_13).

Admin endpoint payload contract for ``GET /api/v2/admin/system-config``
+ ``GET /api/v2/admin/system-config/{key}`` + ``PATCH /api/v2/admin/
system-config/{key}``. Service layer (``SystemConfigService``) owns
the admin-only governance guard.

The ``value`` field is intentionally typed as ``Any`` because the
table holds heterogeneous JSONB shapes (bool / int / list / nested
object). Callers that consume specific keys cast to the expected
shape at use-site; SystemConfigService does NOT validate the inner
shape (admin's responsibility per Q10 governance scope).
"""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class SystemConfigResponse(BaseModel):
    """Single config row response."""

    model_config = ConfigDict(from_attributes=True)

    key: str
    value: Any = Field(
        description=(
            "Config value — any JSON shape (bool, int, list, nested "
            "object). Caller casts to expected type at use-site."
        ),
    )
    description: Optional[str] = None
    updated_at: datetime
    updated_by_user_id: Optional[int] = None


class SystemConfigUpdate(BaseModel):
    """Admin PATCH payload for a single config key.

    ``value`` is REQUIRED — admin always commits a new value. To
    "clear" a key, admin should send an explicit ``null`` /
    ``false`` / ``0`` / ``""`` rather than DELETE the row, so the
    audit row + ``description`` survive.

    ``description`` is optional — admin can edit the docstring
    independently or alongside the value.
    """

    model_config = ConfigDict(extra="forbid")

    value: Any = Field(
        description=(
            "New config value. Any JSON shape; service does not "
            "validate inner structure."
        ),
    )
    description: Optional[str] = Field(
        default=None,
        description=(
            "Optional. Update the admin-facing docstring "
            "alongside the value."
        ),
    )


class SystemConfigListResponse(BaseModel):
    """List response for admin overview."""

    total: int
    items: list[SystemConfigResponse]

File: frontend/src/lib/zod/system-config.ts
/**
 * Zod schemas for system_config admin surface.
 *
 * Mirrors backend ``Backend_FastAPI/app/schemas/system_config.py`` —
 * ``value`` is intentionally untyped (``z.unknown()``) because the
 * table stores heterogeneous JSONB shapes (bool / int / string /
 * list / nested object). The admin UI renders + edits the raw JSON;
 * shape validation belongs at the consumer site (engine reads
 * ``max_choices_per_profile`` as int, storefront reads
 * ``current_intake_year`` as int, etc.).
 *
 * Endpoints:
 *   GET    /api/v2/admin/system-config
 *   GET    /api/v2/admin/system-config/{key}
 *   PATCH  /api/v2/admin/system-config/{key}   (admin-only)
 */
import { z } from "zod"

export const SystemConfigResponseSchema = z.object({
  key: z.string(),
  value: z.unknown(),
  description: z.string().nullable(),
  updated_at: z.string(),
  updated_by_user_id: z.number().int().nullable(),
})

export const SystemConfigUpdateSchema = z.object({
  // ``value`` REQUIRED per BE contract — to "clear" admin sends explicit
  // null/false/0/"" rather than DELETE row. Keep audit + description intact.
  value: z.unknown(),
  description: z.string().nullable().optional(),
})

export const SystemConfigListResponseSchema = z.object({
  total: z.number().int(),
  items: z.array(SystemConfigResponseSchema),
})

export type SystemConfigResponse = z.infer<typeof SystemConfigResponseSchema>
export type SystemConfigUpdate = z.infer<typeof SystemConfigUpdateSchema>
export type SystemConfigListResponse = z.infer<typeof SystemConfigListResponseSchema>
